File: examples/python/rag_chat.py
"""Wire ContIndex into an existing application's RAG chat loop.

The host supplies document retrieval, generation, authorization, stable turn IDs, and
per-history serialization. Index separately with cci.index.index on the host's schedule.
"""
from cci.ingest import ingest
from cci.memory import prepare_context
from cci.models import InputMessage


async def chat_turn(store, question, turn_id, *, retrieve_documents, generate, memory_provider=None):
    # Capture prior context before saving this new user message.
    memory = await prepare_context(store, question, provider=memory_provider)
    await ingest(store, store.history_id, [InputMessage(role="user", content=question)],
                 source_id="rag-chat", idempotency_key=f"{turn_id}:user")
    documents = await retrieve_documents(question)
    # The host's generation adapter keeps trusted instructions separate from both data inputs.
    answer = await generate(question=question, documents=documents, conversation_memory=memory.text)
    await ingest(store, store.history_id, [InputMessage(role="assistant", content=answer)],
                 source_id="rag-chat", idempotency_key=f"{turn_id}:assistant")
    return answer, memory

File: examples/typescript/rag-chat.mjs
/** Adapter for an existing host's document retriever and model. Serialize turns per history. */
import { ingest, prepareContext } from "chat-context-index";

export async function chatTurn(store, question, turnId, { retrieveDocuments, generate, memoryProvider }) {
  const memory = await prepareContext(store, question, { provider: memoryProvider });
  await ingest(store, store.historyId, [{ role: "user", content: question }], "rag-chat", `${turnId}:user`);
  const documents = await retrieveDocuments(question);
  const answer = await generate({ question, documents, conversationMemory: memory.text });
  await ingest(store, store.historyId, [{ role: "assistant", content: answer }], "rag-chat", `${turnId}:assistant`);
  return { answer, memory };
}
